File: OJ practices/LeetCode题目集/92. Reverse Linked List II(medium).java
/*
 * @Author: SourDumplings
 * @Date: 2020-08-21 16:53:04
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode-cn.com/problems/reverse-linked-list-ii/
 */

class Solution
{
    public ListNode reverseBetween(ListNode head, int m, int n)
    {
        if (n - m < 1)
        {
            return head;
        }
        ListNode f = new ListNode(0);
        f.next = head;

        ListNode h = f;
        ListNode t = f;
        ListNode p = h;

        for (int i = 0; i <= n; i++)
        {
            if (i == m - 1)
            {
                h = p;
            }
            if (i == n)
            {
                t = p.next;
            }
            p = p.next;
        }
        p = h.next;
        while (p.next != t)
        {
            ListNode temp = p.next;
            p.next = temp.next;
            temp.next = h.next;
            h.next = temp;
        }
        return f.next;
    }
}

File: OJ practices/LeetCode题目集/92. Reverse Linked List II(medium).cpp
/**
 * @file 92. Reverse Linked List II(medium).cpp
 * @author SourDumplings (sourdumplings@qq.com)
 * @brief https://leetcode.cn/problems/reverse-linked-list-ii/?envType=study-plan-v2&envId=top-interview-150
 * @version 1.0.0
 * @date 2024-11-26
 * 
 * @copyright Copyright (c) 2024 SourDumplings
 * 
 */

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
	ListNode* reverseBetween(ListNode* head, int left, int right)
	{
		ListNode* tempHead = new ListNode(0, head);
		int count		   = 1;
		ListNode* f		   = tempHead;
		ListNode* revHead  = tempHead;

		while (f->next)
		{
			ListNode* p = f->next;
			if (count < left) { f = p; }
			else if (count == left)
			{
				revHead = f;
				f		= p;
			}
			else if (left < count && count <= right)
			{
				f->next		  = p->next;
				p->next		  = revHead->next;
				revHead->next = p;
			}
			else { break; }

			++count;
		}
		ListNode* ret = tempHead->next;
		delete tempHead;
		return ret;
	}
};